feat!(c++): Refactor code structure to consolidate include and src into a single src directory

acezen opened the pull request:
<!--
Thanks for contributing to GraphAr.
If this is your first pull request you can find detailed information on [CONTRIBUTING.md](https://github.com/apache/graphar/blob/main/CONTRIBUTING.md)

The Apache GraphAr (incubating) community has restrictions on the naming of pr title. You can find instructions in
[CONTRIBUTING.md](https://github.com/apache/graphar/blob/main/CONTRIBUTING.md#title) too.
-->

### Reason for this PR
<!-- 
Why are you proposing this change? If this is already tracked in an issue, please link to the issue here.
Explaining clearly why this change is beneficial is important for the reviewers to understand the context.
-->

Our current project structure separates header files and source files into include and src directories, respectively. This approach, while beneficial in some contexts, adds complexity and overhead for smaller projects, **and not fits to protobuf generated code**. To simplify the structure and enhance maintainability, we need to consolidate all code into a single src directory.

### What changes are included in this PR?
<!-- 
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->
Here are some notable changes:
- Consolidate the header and source code into single `src` directory.
- Add subdirectory to classify the `arrow` and `high-level` related code.
- Add `api` headers to let users to easily to include the function of c++ library.

### Are these changes tested?
<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->
Yes

### Are there any user-facing changes?
<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please uncomment the line below and explain which changes are breaking.
-->
**BREAKING CHANGE:**
- the header path and header name may change, like:
  - `graphar/arrow_chunk_reader.h` to `graphar/arrow/chunk_reader.h`
  - `graphar/arrow_chunk_writer.h` to `graphar/arrow/chunk_writer.h`
- The `copy` way writer function has been moved from `arrow_chunk_writer.h` to `chunk_info_writer.h`

<!--
Please uncomment the line below (and provide explanation) if the changes fix either
(a) a security vulnerability,
(b) a bug that caused incorrect or invalid data to be produced, or
(c) a bug that causes a crash (even when the API contract is upheld). 
We use this to highlight fixes to issues that may affect users without their knowledge. For this reason, fixing bugs that cause errors don't count, since those are usually obvious.
-->
<!-- **Critical Fix: <description>** -->

acezen commented:
Hi, @lixueclaire, I create this PR to consolidate all code into a single src directory and brings some `BREAKING CHANGE`, may need your opinion about the changes, especially the naming of the directory and file:
- I use `high-level` for our high level reader and writer directory name
- Since we use `arrow` folder to separate the arrow reader and writer, so I change the `arrow_chunk_reader` and `arrow_chunk_writer` to `chunk_reader` and `chunk_writer`.
- And also the `api` 's header names.

lixueclaire commented:
> Hi, @lixueclaire, I create this PR to consolidate all code into a single src directory and brings some `BREAKING CHANGE`, may need your opinion about the changes, especially the naming of the directory and file:
> 
> * I use `high-level` for our high level reader and writer directory name
> * Since we use `arrow` folder to separate the arrow reader and writer, so I change the `arrow_chunk_reader` and `arrow_chunk_writer` to `chunk_reader` and `chunk_writer`.
> * And also the `api` 's header names.

Hi, I think the APIs maybe a little confusing now. Originally, we have `chunk_info_reader` and `arrow_chunk_reader` and high-level graph collections. If we use `basic_reader`, it is not clear what is read. I suggest naming `info_reader/meta_reader`,`arrow_reader/chunk_reader` and `high_level_reader/graph_reader`.

acezen commented:
> > Hi, @lixueclaire, I create this PR to consolidate all code into a single src directory and brings some `BREAKING CHANGE`, may need your opinion about the changes, especially the naming of the directory and file:
> > 
> > * I use `high-level` for our high level reader and writer directory name
> > * Since we use `arrow` folder to separate the arrow reader and writer, so I change the `arrow_chunk_reader` and `arrow_chunk_writer` to `chunk_reader` and `chunk_writer`.
> > * And also the `api` 's header names.
> 
> Hi, I think the APIs maybe a little confusing now. Originally, we have `chunk_info_reader` and `arrow_chunk_reader` and high-level graph collections. If we use `basic_reader`, it is not clear what is read. I suggest naming `info_reader/meta_reader`,`arrow_reader/chunk_reader` and `high_level_reader/graph_reader`.

You means we should split reader and writer to `info`, `arrow` and `high_level`and also split the same level reader/writer to different directory?

lixueclaire commented:
> > > Hi, @lixueclaire, I create this PR to consolidate all code into a single src directory and brings some `BREAKING CHANGE`, may need your opinion about the changes, especially the naming of the directory and file:
> > > 
> > > * I use `high-level` for our high level reader and writer directory name
> > > * Since we use `arrow` folder to separate the arrow reader and writer, so I change the `arrow_chunk_reader` and `arrow_chunk_writer` to `chunk_reader` and `chunk_writer`.
> > > * And also the `api` 's header names.
> > 
> > 
> > Hi, I think the APIs maybe a little confusing now. Originally, we have `chunk_info_reader` and `arrow_chunk_reader` and high-level graph collections. If we use `basic_reader`, it is not clear what is read. I suggest naming `info_reader/meta_reader`,`arrow_reader/chunk_reader` and `high_level_reader/graph_reader`.
> 
> You mean we should split reader and writer to `info`, `arrow` and `high_level`and also split the same level reader/writer to different directory?

I'm talking about the file names under the api directory.

acezen commented:
> > > > Hi, @lixueclaire, I create this PR to consolidate all code into a single src directory and brings some `BREAKING CHANGE`, may need your opinion about the changes, especially the naming of the directory and file:
> > > > 
> > > > * I use `high-level` for our high level reader and writer directory name
> > > > * Since we use `arrow` folder to separate the arrow reader and writer, so I change the `arrow_chunk_reader` and `arrow_chunk_writer` to `chunk_reader` and `chunk_writer`.
> > > > * And also the `api` 's header names.
> > > 
> > > 
> > > Hi, I think the APIs maybe a little confusing now. Originally, we have `chunk_info_reader` and `arrow_chunk_reader` and high-level graph collections. If we use `basic_reader`, it is not clear what is read. I suggest naming `info_reader/meta_reader`,`arrow_reader/chunk_reader` and `high_level_reader/graph_reader`.
> > 
> > 
> > You mean we should split reader and writer to `info`, `arrow` and `high_level`and also split the same level reader/writer to different directory?
> 
> I'm talking about the file names under the api directory.

I have changed the naming under api, could you help reviewing again?

lixueclaire commented on cpp/examples/low_level_reader_example.cc:
```diff
@@ -23,8 +23,7 @@
 #include "./config.h"
-#include "graphar/api.h"
-#include "graphar/reader/chunk_info_reader.h"
+#include "graphar/api/basic_reader.h"
```
This need to be updated.

acezen replied:
Thanks for pointing out, fixed.

lixueclaire reviewed (approved):
LGTM now. And it's essential to update our documentation accordingly.
